app/models/manageiq/providers/ibm_cloud/power_virtual_servers/cloud_manager/event_catcher/stream.rb

```diff
@@ -17,32 +15,21 @@ def stop
     @stop_polling = true
   end
 
-  def fake_events
-    [
-      OpenStruct.new(:name       => %w[instance_power_on instance_power_off].sample,
-                     :id         => Time.zone.now.to_i,
-                     :timestamp  => Time.zone.now,
-                     :vm_ems_ref => [1, 2].sample),
-      OpenStruct.new(:name       => %w[instance_power_on instance_power_off].sample,
-                     :id         => Time.zone.now.to_i + 1,
-                     :timestamp  => Time.zone.now,
-                     :vm_ems_ref => [1, 2].sample),
-    ]
-  end
-
   def poll
-    @ems.with_provider_connection do |_connection|
-      catch(:stop_polling) do
-        loop do
-          fake_events.each do |activity|
-            throw :stop_polling if @stop_polling
-            yield activity.to_h
-          end
-          sleep @poll_sleep
-        end
-      rescue => exception
-        raise ProviderUnreachable, exception.message
+    from_time = Time.now.utc.to_i
+
+    pcloud_events_api = ems.connect(:service => "PCloudEventsApi")
+    loop do
+      begin
+        events = pcloud_events_api.pcloud_events_getsince(@ems.uid_ems, from_time).events
+      # TODO: Only rescue token experation exception
+      rescue Exception => e
+        pcloud_events_api = ems.connect(:service => "PCloudEventsApi")
       end
+      from_time = Time.now.utc.to_i
+      events.each { |event| yield event.to_hash }
+      break if stop_polling
+      sleep(poll_sleep)
```

Comment:
You can use `retry` and it'll go back to the top of the current block, should shave you having to call ems.connect in two different places.

Try:
```suggestion
    loop do
      pcloud_events_api = ems.connect(:service => "PCloudEventsApi")
      
      retry_connection = true

      events = pcloud_events_api.pcloud_events_getsince(@ems.uid_ems, from_time).events
      
      from_time = Time.now.utc.to_i
      events.each { |event| yield event.to_hash }
      break if stop_polling
      sleep(poll_sleep)
    rescue IbmCloudPower::ApiError => e
      raise unless e.code == 403 && retry_connection
      retry_connection = false
      retry
    end
```